File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;
class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;

        HashMap<String, Integer> map = new HashMap<>();
        for (int i=0;i<want.length;i++){
            map.put(want[i], number[i]);
        }
        // discount 돌면서
        // 1. left = 0, right = 9로 설정하고, 
        // 2. 돌면서 map 안에 있으면 -1 하기
        // 3. right를 discount.length 까지 돌면서 map의 value가 모두 1보다 작으면 answer++
        // 초기 세팅
        int left = 0;
        int right = 9;
        for (int i = left;i<=right;i++){
                String target = discount[i];
                if (map.containsKey(target)){
                    map.put(target, map.get(target) - 1);
                } 
        }
        boolean flag = true;
        for(String s : want) {
            if (map.get(s) >= 1) {
                flag = false;
                break;
            }
        }
        if (flag) answer++;
    
        while (right < discount.length - 1) {
            
            String left_val = discount[left++];
            
            String right_val = discount[++right];
            
            
            if (map.containsKey(left_val)) map.put(left_val, map.get(left_val) + 1);
            if (map.containsKey(right_val)) map.put(right_val, map.get(right_val) - 1);
            
            flag = true;
            for(String s : want) {
                if (map.get(s) >= 1) {
                    flag = false;
                    break;
                }
            }
            if (flag) answer++;
        }
    
        
        
        return answer;
    }
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.cpp
#include <bits/stdc++.h>

using namespace std;

bool checkMap(map<string, int> &m) {
    for (auto const& [key, val] : m) {
        if (val > 0) { // 아직 필요한 물건이 남아있다면
            return false; // 실패
        }
    }
    return true; // 모든 물건의 필요 수량이 0 이하가 되었으면 성공
}

int solution(vector<string> want, vector<int> number, vector<string> discount) {
    int answer = 0;
    map<string, int> m;

    for (int i=0;i<want.size();i++) {
        m[want[i]] = number[i];
    }
    // 0 ~ 9까지 체크
    for (int i=0;i<10;i++) {
        if (m.count(discount[i])) m[discount[i]]--;
    }

    if (checkMap(m)) answer++;

    // 슬라이딩 윈도우
    for (int start = 0, end = 10; end < discount.size(); start++, end++) {
        // 빠지는 상품 복구
        if (m.count(discount[start])) {
            m[discount[start]]++;
        }

        // 들어오는 상품 감소
        if (m.count(discount[end])) {
            m[discount[end]]--;
        }

        if (checkMap(m)) answer++;
    }




    return answer;
}